Bottom: 01884644a9dd0808c72d178121523931a801dc07
Top:    1b6f46bd81a3d1093d292803e60903de15f4adf0
Author: Peter Clifton <peter@clifton-electronics.co.uk>
Date:   2012-12-24 02:14:27 +0000

find.c: Avoid modifying bounds of the objects we are passing down

Bloat the bounding box into a local BoxType structure instead, to
avoid passing inconsistent objects to lower level functions.


---

diff --git a/src/find.c b/src/find.c
index dae0a11..7e93380 100644
--- a/src/find.c
+++ b/src/find.c
@@ -104,12 +104,6 @@
  * some local macros
  */
 
-#define EXPAND_BOUNDS(p) if (Bloat > 0) {\
-       (p)->BoundingBox.X1 -= Bloat; \
-       (p)->BoundingBox.X2 += Bloat; \
-       (p)->BoundingBox.Y1 -= Bloat; \
-       (p)->BoundingBox.Y2 += Bloat;}
-
 #define	SEPARATE(FP)							\
 	{											\
 		int	i;									\
@@ -399,6 +393,22 @@ ADD_POLYGON_TO_LIST (Cardinal L, PolygonType *Ptr)
   return add_object_to_list (&PolygonList[L], POLYGON_TYPE, LAYER_PTR (L), Ptr, Ptr);
 }
 
+static BoxType
+expand_bounds (BoxType *box_in)
+{
+  BoxType box_out = *box_in;
+
+  if (Bloat > 0)
+    {
+      box_out.X1 -= Bloat;
+      box_out.X2 += Bloat;
+      box_out.Y1 -= Bloat;
+      box_out.Y2 += Bloat;
+    }
+
+  return box_out;
+}
+
 bool
 PinLineIntersect (PinType *PV, LineType *Line)
 {
@@ -729,13 +739,15 @@ LookupLOConnectionsToPVList (bool AndRats)
   /* loop over all PVs currently on list */
   while (PVList.Location < PVList.Number)
     {
+      BoxType search_box;
+
       /* get pointer to data */
       info.pv = *(PVLIST_ENTRY (PVList.Location));
-      EXPAND_BOUNDS (&info.pv);
+      search_box = expand_bounds (&info.pv.BoundingBox);
 
       /* check pads */
       if (setjmp (info.env) == 0)
-        r_search (PCB->Data->pad_tree, (BoxType *) & info.pv, NULL,
+        r_search (PCB->Data->pad_tree, &search_box, NULL,
                   LOCtoPVpad_callback, &info);
       else
         return true;
@@ -749,21 +761,22 @@ LookupLOConnectionsToPVList (bool AndRats)
              continue;
 
           info.layer = layer_no;
+
           /* add touching lines */
           if (setjmp (info.env) == 0)
-            r_search (layer->line_tree, (BoxType *) & info.pv,
+            r_search (layer->line_tree, &search_box,
                       NULL, LOCtoPVline_callback, &info);
           else
             return true;
           /* add touching arcs */
           if (setjmp (info.env) == 0)
-            r_search (layer->arc_tree, (BoxType *) & info.pv,
+            r_search (layer->arc_tree, &search_box,
                       NULL, LOCtoPVarc_callback, &info);
           else
             return true;
           /* check all polygons */
           if (setjmp (info.env) == 0)
-            r_search (layer->polygon_tree, (BoxType *) & info.pv,
+            r_search (layer->polygon_tree, &search_box,
                       NULL, LOCtoPVpoly_callback, &info);
           else
             return true;
@@ -772,7 +785,7 @@ LookupLOConnectionsToPVList (bool AndRats)
       if (AndRats)
         {
           if (setjmp (info.env) == 0)
-            r_search (PCB->Data->rat_tree, (BoxType *) & info.pv, NULL,
+            r_search (PCB->Data->rat_tree, &search_box, NULL,
                       LOCtoPVrat_callback, &info);
           else
             return true;
@@ -936,16 +949,19 @@ LookupPVConnectionsToPVList (void)
   save_place = PVList.Location;
   while (PVList.Location < PVList.Number)
     {
+      BoxType search_box;
+
       /* get pointer to data */
       info.pv = *(PVLIST_ENTRY (PVList.Location));
-      EXPAND_BOUNDS (&info.pv);
+      search_box = expand_bounds ((BoxType *)&info.pv);
+
       if (setjmp (info.env) == 0)
-        r_search (PCB->Data->via_tree, (BoxType *) & info.pv, NULL,
+        r_search (PCB->Data->via_tree, &search_box, NULL,
                   pv_pv_callback, &info);
       else
         return true;
       if (setjmp (info.env) == 0)
-        r_search (PCB->Data->pin_tree, (BoxType *) & info.pv, NULL,
+        r_search (PCB->Data->pin_tree, &search_box, NULL,
                   pv_pv_callback, &info);
       else
         return true;
@@ -1107,15 +1123,18 @@ LookupPVConnectionsToLOList (bool AndRats)
       /* check all lines */
       while (LineList[layer_no].Location < LineList[layer_no].Number)
         {
+          BoxType search_box;
+
           info.line = *(LINELIST_ENTRY (layer_no, LineList[layer_no].Location));
-          EXPAND_BOUNDS (&info.line);
+          search_box = expand_bounds ((BoxType *)&info.line);
+
           if (setjmp (info.env) == 0)
-            r_search (PCB->Data->via_tree, (BoxType *) & info.line, NULL,
+            r_search (PCB->Data->via_tree, &search_box, NULL,
                       pv_line_callback, &info);
           else
             return true;
           if (setjmp (info.env) == 0)
-            r_search (PCB->Data->pin_tree, (BoxType *) & info.line, NULL,
+            r_search (PCB->Data->pin_tree, &search_box, NULL,
                       pv_line_callback, &info);
           else
             return true;
@@ -1125,15 +1144,18 @@ LookupPVConnectionsToLOList (bool AndRats)
       /* check all arcs */
       while (ArcList[layer_no].Location < ArcList[layer_no].Number)
         {
+          BoxType search_box;
+
           info.arc = *(ARCLIST_ENTRY (layer_no, ArcList[layer_no].Location));
-          EXPAND_BOUNDS (&info.arc);
+          search_box = expand_bounds ((BoxType *)&info.arc);
+
           if (setjmp (info.env) == 0)
-            r_search (PCB->Data->via_tree, (BoxType *) & info.arc, NULL,
+            r_search (PCB->Data->via_tree, &search_box, NULL,
                       pv_arc_callback, &info);
           else
             return true;
           if (setjmp (info.env) == 0)
-            r_search (PCB->Data->pin_tree, (BoxType *) & info.arc, NULL,
+            r_search (PCB->Data->pin_tree, &search_box, NULL,
                       pv_arc_callback, &info);
           else
             return true;
@@ -1144,16 +1166,18 @@ LookupPVConnectionsToLOList (bool AndRats)
       info.layer = layer_no;
       while (PolygonList[layer_no].Location < PolygonList[layer_no].Number)
         {
-          info.polygon =
-            *(POLYGONLIST_ENTRY (layer_no, PolygonList[layer_no].Location));
-          EXPAND_BOUNDS (&info.polygon);
+          BoxType search_box;
+
+          info.polygon = *(POLYGONLIST_ENTRY (layer_no, PolygonList[layer_no].Location));
+          search_box = expand_bounds ((BoxType *)&info.polygon);
+
           if (setjmp (info.env) == 0)
-            r_search (PCB->Data->via_tree, (BoxType *) & info.polygon, NULL,
+            r_search (PCB->Data->via_tree, &search_box, NULL,
                       pv_poly_callback, &info);
           else
             return true;
           if (setjmp (info.env) == 0)
-            r_search (PCB->Data->pin_tree, (BoxType *) & info.polygon, NULL,
+            r_search (PCB->Data->pin_tree, &search_box, NULL,
                       pv_poly_callback, &info);
           else
             return true;
@@ -1176,15 +1200,18 @@ LookupPVConnectionsToLOList (bool AndRats)
        */
       while (PadList[layer_no].Location < PadList[layer_no].Number)
         {
+          BoxType search_box;
+
           info.pad = *(PADLIST_ENTRY (layer_no, PadList[layer_no].Location));
-          EXPAND_BOUNDS (&info.pad);
+          search_box = expand_bounds ((BoxType *)&info.pad);
+
           if (setjmp (info.env) == 0)
-            r_search (PCB->Data->via_tree, (BoxType *) & info.pad, NULL,
+            r_search (PCB->Data->via_tree, &search_box, NULL,
                       pv_pad_callback, &info);
           else
             return true;
           if (setjmp (info.env) == 0)
-            r_search (PCB->Data->pin_tree, (BoxType *) & info.pad, NULL,
+            r_search (PCB->Data->pin_tree, &search_box, NULL,
                       pv_pad_callback, &info);
           else
             return true;
@@ -1705,9 +1732,11 @@ LookupLOConnectionsToArc (ArcType *Arc, Cardinal LayerGroup)
 {
   Cardinal entry;
   struct lo_info info;
+  BoxType search_box;
 
   info.arc = *Arc;
-  EXPAND_BOUNDS (&info.arc);
+  search_box = expand_bounds ((BoxType *)&info.arc);
+
   /* loop over all layers of the group */
   for (entry = 0; entry < PCB->LayerGroups.Number[LayerGroup]; entry++)
     {
@@ -1724,13 +1753,13 @@ LookupLOConnectionsToArc (ArcType *Arc, Cardinal LayerGroup)
           info.layer = layer_no;
           /* add arcs */
           if (setjmp (info.env) == 0)
-            r_search (layer->line_tree, &info.arc.BoundingBox,
+            r_search (layer->line_tree, &search_box,
                       NULL, LOCtoArcLine_callback, &info);
           else
             return true;
 
           if (setjmp (info.env) == 0)
-            r_search (layer->arc_tree, &info.arc.BoundingBox,
+            r_search (layer->arc_tree, &search_box,
                       NULL, LOCtoArcArc_callback, &info);
           else
             return true;
@@ -1748,7 +1777,7 @@ LookupLOConnectionsToArc (ArcType *Arc, Cardinal LayerGroup)
         {
           info.layer = layer_no - max_copper_layer;
           if (setjmp (info.env) == 0)
-            r_search (PCB->Data->pad_tree, &info.arc.BoundingBox, NULL,
+            r_search (PCB->Data->pad_tree, &search_box, NULL,
                       LOCtoArcPad_callback, &info);
           else
             return true;
@@ -1837,13 +1866,15 @@ LookupLOConnectionsToLine (LineType *Line, Cardinal LayerGroup,
 {
   Cardinal entry;
   struct lo_info info;
+  BoxType search_box;
 
-  info.line = *Line;
   info.layer = LayerGroup;
-  EXPAND_BOUNDS (&info.line)
+  info.line = *Line;
+  search_box = expand_bounds ((BoxType *)&info.line);
+
   /* add the new rat lines */
   if (setjmp (info.env) == 0)
-    r_search (PCB->Data->rat_tree, &info.line.BoundingBox, NULL,
+    r_search (PCB->Data->rat_tree, &search_box, NULL,
               LOCtoLineRat_callback, &info);
   else
     return true;
@@ -1863,13 +1894,13 @@ LookupLOConnectionsToLine (LineType *Line, Cardinal LayerGroup,
           info.layer = layer_no;
           /* add lines */
           if (setjmp (info.env) == 0)
-            r_search (layer->line_tree, (BoxType *) & info.line,
+            r_search (layer->line_tree, &search_box,
                       NULL, LOCtoLineLine_callback, &info);
           else
             return true;
           /* add arcs */
           if (setjmp (info.env) == 0)
-            r_search (layer->arc_tree, (BoxType *) & info.line,
+            r_search (layer->arc_tree, &search_box,
                       NULL, LOCtoLineArc_callback, &info);
           else
             return true;
@@ -1892,7 +1923,7 @@ LookupLOConnectionsToLine (LineType *Line, Cardinal LayerGroup,
           /* handle special 'pad' layers */
           info.layer = layer_no - max_copper_layer;
           if (setjmp (info.env) == 0)
-            r_search (PCB->Data->pad_tree, &info.line.BoundingBox, NULL,
+            r_search (PCB->Data->pad_tree, &search_box, NULL,
                       LOCtoLinePad_callback, &info);
           else
             return true;
@@ -2109,16 +2140,19 @@ LookupLOConnectionsToPad (PadType *Pad, Cardinal LayerGroup)
 {
   Cardinal entry;
   struct lo_info info;
+  BoxType search_box;
 
   if (!TEST_FLAG (SQUAREFLAG, Pad))
     return (LookupLOConnectionsToLine ((LineType *) Pad, LayerGroup, false));
 
   info.pad = *Pad;
-  EXPAND_BOUNDS (&info.pad);
+  search_box = expand_bounds ((BoxType *)&info.pad);
+
   /* add the new rat lines */
   info.layer = LayerGroup;
+
   if (setjmp (info.env) == 0)
-    r_search (PCB->Data->rat_tree, &info.pad.BoundingBox, NULL,
+    r_search (PCB->Data->rat_tree, &search_box, NULL,
               LOCtoPadRat_callback, &info);
   else
     return true;
@@ -2137,19 +2171,19 @@ LookupLOConnectionsToPad (PadType *Pad, Cardinal LayerGroup)
           info.layer = layer_no;
           /* add lines */
           if (setjmp (info.env) == 0)
-            r_search (layer->line_tree, &info.pad.BoundingBox,
+            r_search (layer->line_tree, &search_box,
                       NULL, LOCtoPadLine_callback, &info);
           else
             return true;
           /* add arcs */
           if (setjmp (info.env) == 0)
-            r_search (layer->arc_tree, &info.pad.BoundingBox,
+            r_search (layer->arc_tree, &search_box,
                       NULL, LOCtoPadArc_callback, &info);
           else
             return true;
           /* add polygons */
           if (setjmp (info.env) == 0)
-            r_search (layer->polygon_tree, &info.pad.BoundingBox,
+            r_search (layer->polygon_tree, &search_box,
                       NULL, LOCtoPadPoly_callback, &info);
           else
             return true;
@@ -2159,7 +2193,7 @@ LookupLOConnectionsToPad (PadType *Pad, Cardinal LayerGroup)
           /* handle special 'pad' layers */
           info.layer = layer_no - max_copper_layer;
           if (setjmp (info.env) == 0)
-            r_search (PCB->Data->pad_tree, (BoxType *) & info.pad, NULL,
+            r_search (PCB->Data->pad_tree, &search_box, NULL,
                       LOCtoPadPad_callback, &info);
           else
             return true;
@@ -2245,18 +2279,21 @@ LookupLOConnectionsToPolygon (PolygonType *Polygon, Cardinal LayerGroup)
 {
   Cardinal entry;
   struct lo_info info;
+  BoxType search_box;
 
   if (!Polygon->Clipped)
     return false;
   info.polygon = *Polygon;
-  EXPAND_BOUNDS (&info.polygon);
+  search_box = expand_bounds ((BoxType *)&info.polygon);
+
   info.layer = LayerGroup;
   /* check rats */
   if (setjmp (info.env) == 0)
-    r_search (PCB->Data->rat_tree, (BoxType *) & info.polygon, NULL,
+    r_search (PCB->Data->rat_tree, &search_box, NULL,
               LOCtoPolyRat_callback, &info);
   else
     return true;
+
 /* loop over all layers of the group */
   for (entry = 0; entry < PCB->LayerGroups.Number[LayerGroup]; entry++)
     {
@@ -2284,14 +2321,13 @@ LookupLOConnectionsToPolygon (PolygonType *Polygon, Cardinal LayerGroup)
           info.layer = layer_no;
           /* check all lines */
           if (setjmp (info.env) == 0)
-            r_search (layer->line_tree,
-                      (BoxType *) & info.polygon, NULL,
-                      LOCtoPolyLine_callback, &info);
+            r_search (layer->line_tree, &search_box,
+                      NULL, LOCtoPolyLine_callback, &info);
           else
             return true;
           /* check all arcs */
           if (setjmp (info.env) == 0)
-            r_search (layer->arc_tree, (BoxType *) & info.polygon,
+            r_search (layer->arc_tree, &search_box,
                       NULL, LOCtoPolyArc_callback, &info);
           else
             return true;
@@ -2300,7 +2336,7 @@ LookupLOConnectionsToPolygon (PolygonType *Polygon, Cardinal LayerGroup)
         {
           info.layer = layer_no - max_copper_layer;
           if (setjmp (info.env) == 0)
-            r_search (PCB->Data->pad_tree, (BoxType *) & info.polygon,
+            r_search (PCB->Data->pad_tree, &search_box,
                       NULL, LOCtoPolyPad_callback, &info);
           else
             return true;
